[Issue1379] convert number predicates to boolean

xysun opened the pull request:
by checking whether they are equal to zero

Fixes #1379

xysun committed:
convert number predicates to boolean by checking whether they are equal to zero

xysun commented:
@shoyer Please take a look. Thanks!

xysun commented on jax/lax/lax_control_flow.py:
```diff
@@ -287,6 +288,13 @@ def cond(pred, true_operand, true_fun, false_operand, false_fun):
   """
+  pred_dtype = array(pred).dtype
+  if pred_dtype.kind != 'b':
+    if pred_dtype.kind in 'iuf':
```
allow int, unsigned int, float
(maybe too much happy to narrow down)

shoyer replied:
I think "anything that's okay in `np.where`" is probably a good guideline here. `np.where` works with all numeric types, so this looks good to me.

xysun committed:
one more test

xysun committed:
try relative import for python 2.7

shoyer commented on tests/lax_control_flow_test.py:
```diff
@@ -416,6 +416,34 @@ def false_fun(x):
 
+  def testIssue1379(self):
+    # integer predicates should be treated as boolean by testing whether they are zero
+    def fun(x):
```
I'm not quite sure what the point of this function is -- it looks like it's testing Python itself?

xysun replied:
This is used to test that Python and jax create same outputs.

xysun replied:
I have updated tests. Now the tests are testing regular lax vs jit compiled lax function calls.

xysun committed:
try fixing circular import in python 2.7 take 2

shoyer commented on tests/lax_control_flow_test.py:
```diff
@@ -416,6 +416,34 @@ def false_fun(x):
+    # test float predicates
+    @api.jit
+    def cfun_float(x):
+      return lax.cond(lax.rem(x, 2.), x, lambda x: x, x, lambda x: lax.add(x, 1.))
```
Rather than using the `lax.rem()` operator internally, could you simply this to pass in the `pred` value directly?

shoyer replied:
Also, you might be able to get away with same helper function for both integers and floats.

shoyer commented on tests/lax_control_flow_test.py:
```diff
@@ -416,6 +416,34 @@ def false_fun(x):
+    self.assertEqual(cfun_float(2.5), 2.5)
+    self.assertEqual(fun(2.), cfun_float(2.))
+    self.assertEqual(cfun_float(2.), 3.)
+    
```
It would be nice to add a test for the error message if you pass in an invalid dtype.

xysun committed:
simplify test; add test for wrong type

xysun commented:
@shoyer In fc632b2 I simplified the test also added test for the error, now trying to fix circular import problem for Python 2.7...

xysun committed:
more comprehensive tests; block collection type

shoyer commented:
To avoid the recursive import, you can use onp.shape and onp.result_type

xysun committed:
use numpy to avoid recursive imports

xysun commented:
Thanks @shoyer ! 
I have updated to use numpy; also added tests for both regular and jit compiled functions. 
Also put an explicit check against collection types (although Python conditional allows predicates like lists/tuple/dict, in this PR I thought it's better to limit to numerical values only)

shoyer commented:
thanks @xysun !
